internal/v1/handler_blueprints.go

```diff
@@ -33,12 +34,45 @@ type BlueprintBody struct {
 	ImageRequests  []ImageRequest `json:"image_requests"`
 }
 
-func BlueprintFromAPI(cbr CreateBlueprintRequest) BlueprintBody {
-	return BlueprintBody{
+func (u *User) CryptPassword() error {
+	// Prevent empty and already hashed password  from being hashed
+	if u.Password == nil || (len(*u.Password) == 0 || crypt.PasswordIsCrypted(*u.Password)) {
+		return nil
+	}
+	pw, err := crypt.CryptSHA512(*u.Password)
+	if err != nil {
+		return err
+	}
+	*u.Password = pw
+	return nil
+}
+
+func (bb *BlueprintBody) CryptPasswords() error {
+	if bb.Customizations.Users != nil {
+		var usersWithHashedPasswords []User
+		for _, u := range *bb.Customizations.Users {
+			err := u.CryptPassword()
+			if err != nil {
+				return err
+			}
+			usersWithHashedPasswords = append(usersWithHashedPasswords, u)
+		}
+		bb.Customizations.Users = &usersWithHashedPasswords
```

Comment:
The `usersWithHashedPasswords` is I assume a workaround around the fact that you could not see original values being changed. This is a [common problem](https://www.calhoun.io/does-range-copy-the-slice-in-go/) and more convenient way to fix that is this:

```
for i := range slice {
	slice[i].CallMethod()
}
```

Or was there other erason to copy all users?

Reply:
Yeah, it was supposed to be just a workaround, interesting, thanks for the article, I had something like that there initially but then the unit tests were failing, I probably just messed it up somehow :shrug: works now with the methods I have :+1: